tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1305 @@
+    ///   does not exist. Presumably the server is not up.
+    /// * [ERROR_PIPE_BUSY] - which needs to be tested for through a constant in
+    ///   [winapi]. This error is raised when the named pipe has been created,
+    ///   but the server is not currently waiting for a connection.
+    ///
+    /// [ERROR_PIPE_BUSY]: crate::winapi::shared::winerror::ERROR_PIPE_BUSY
+    /// [I/O enabled]: crate::runtime::Builder::enable_io
+    /// [Tokio Runtime]: crate::runtime::Runtime
+    /// [winapi]: crate::winapi
+    ///
+    /// A connect loop that waits until a socket becomes available looks like
+    /// this:
+    ///
+    /// ```no_run
+    /// use std::time::Duration;
+    /// use tokio::net::windows::NamedPipeClientOptions;
+    /// use tokio::time;
+    /// use winapi::shared::winerror;
+    ///
+    /// const PIPE_NAME: &str = r"\\.\pipe\mynamedpipe";
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let client = loop {
+    ///     match NamedPipeClientOptions::new().create(PIPE_NAME) {
+    ///         Ok(client) => break client,
+    ///         Err(e) if e.raw_os_error() == Some(winerror::ERROR_PIPE_BUSY as i32) => (),
+    ///         Err(e) => return Err(e),
+    ///     }
+    ///
+    ///     time::sleep(Duration::from_millis(50)).await;
+    /// };
+    ///
+    /// // use the connected client.
+    /// # Ok(()) }
+    /// ```
+    pub fn create(&self, addr: impl AsRef<OsStr>) -> io::Result<NamedPipe> {
+        // Safety: We're calling create_with_security_attributes w/ a null
+        // pointer which disables it.
+        unsafe { self.create_with_security_attributes(addr, ptr::null_mut()) }
+    }
+
+    /// Open the named pipe identified by `addr`.
+    ///
+    /// This is the same as [create][NamedPipeClientOptions::create] except that
+    /// it supports providing security attributes.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `attrs` points to an initialized instance
+    /// of a [SECURITY_ATTRIBUTES] structure.
+    ///
+    /// [SECURITY_ATTRIBUTES]: [crate::winapi::um::minwinbase::SECURITY_ATTRIBUTES]
+    pub unsafe fn create_with_security_attributes(
+        &self,
+        addr: impl AsRef<OsStr>,
+        attrs: *mut (),
+    ) -> io::Result<NamedPipe> {
+        let addr = encode_addr(addr);
+
+        // NB: We could use a platform specialized `OpenOptions` here, but since
+        // we have access to winapi it ultimately doesn't hurt to use
+        // `CreateFile` explicitly since it allows the use of our already
+        // well-structured wide `addr` to pass into CreateFileW.
+        let h = fileapi::CreateFileW(
+            addr.as_ptr(),
+            self.desired_access,
+            0,
+            attrs as *mut _,
+            fileapi::OPEN_EXISTING,
+            winbase::FILE_FLAG_OVERLAPPED | winbase::SECURITY_IDENTIFICATION,
+            ptr::null_mut(),
```

Comment:
Should we provide an option to change `SECURITY_IDENTIFICATION` to something else?

Reply:
We could mimic [`security_qos_flags`](https://doc.rust-lang.org/std/os/windows/fs/trait.OpenOptionsExt.html#tymethod.security_qos_flags).

Feels like we're slowly building our own `OpenOptions` by now with the difference in that ours:
* Ensures FILE_FLAG_OVERLAPPED is set
* Can set `SECURITY_ATTRIBUTES` through `create_with_security_attributes`.

Reply:
I mean, yes, this is basically our own `OpenOptions`. That's fine though.

Reply:
We now have a `security_qos_flags` which is biased slightly towards creating named pipes by having `SECURITY_IDENTIFICATION` be set by default.

This also ensures that `SECURITY_SQOS_PRESENT` is always set (and not just when specified) to ensure that `SECURITY_ANONYMOUS` is respected (see: <https://github.com/rust-lang/rust/pull/58216>).

Now, we could do like `OpenOptions` and allow them to be fully disabled, I personally don't know of any instances where that is desirable for named pipes in production. If someone raises one in the future, we can add another option to disable them then.